cmd/boot-aws/main.go

```diff
@@ -0,0 +1,494 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/google/uuid"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/osbuild/images/internal/cloud/awscloud"
+)
+
+// exitCheck can be deferred from the top of command functions to exit with an
+// error code after any other defers are run in the same scope.
+func exitCheck(err error) {
+	if err != nil {
+		fmt.Fprint(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+// createUserData creates cloud-init's user-data that contains user redhat with
+// the specified public key
+func createUserData(username, publicKeyFile string) (string, error) {
+	publicKey, err := os.ReadFile(publicKeyFile)
+	if err != nil {
+		return "", err
+	}
+
+	userData := fmt.Sprintf(`#cloud-config
+user: %s
+ssh_authorized_keys:
+  - %s
+`, username, string(publicKey))
+
+	return userData, nil
+}
+
+// resources created or allocated for an instance that can be cleaned up when
+// tearing down.
+type resources struct {
+	AMI           *string `json:"ami,omitempty"`
+	Snapshot      *string `json:"snapshot,omitempty"`
+	SecurityGroup *string `json:"security-group,omitempty"`
+	InstanceID    *string `json:"instance,omitempty"`
+}
+
+func run(c string, args ...string) ([]byte, []byte, error) {
+	fmt.Printf("> %s %s\n", c, strings.Join(args, " "))
+	cmd := exec.Command(c, args...)
+
+	var cmdout, cmderr bytes.Buffer
+	cmd.Stdout = &cmdout
+	cmd.Stderr = &cmderr
+	err := cmd.Run()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	stdout := cmdout.Bytes()
+	if len(stdout) > 0 {
+		fmt.Println(string(stdout))
+	}
+
+	stderr := cmderr.Bytes()
+	if len(stderr) > 0 {
+		fmt.Fprintf(os.Stderr, string(stderr)+"\n")
+	}
+	return stdout, stderr, err
+}
+
+func sshRun(ip, user, key, hostsfile string, command ...string) error {
+	sshargs := []string{"-i", key, "-o", fmt.Sprintf("UserKnownHostsFile=%s", hostsfile), "-l", user, ip}
+	sshargs = append(sshargs, command...)
+	_, _, err := run("ssh", sshargs...)
```

Comment:
Would it make sense to print the `stdout` when running the "check" script on the remote host? :thinking:

Reply:
It should be printed by the `run()` function (L70-72).  I'm not sure what's happening, but I think it might be some weirdness with stdout buffers.  The same log you linked prints
```
Testing image at ./build/rhel_9-x86_64-ami-empty/image/image.raw
✅ Marking boot successful
```
after the boot test is finished, but the `Testing image at` line should be printed before the boot test is triggered.

Reply:
I saw that the check script output should be printed in the logs, but I can't see the output of the commands run in it (such as `rpm -qa`, etc...). The reason is most probably still the fact that the script was not yet run in the CI :confused: